### PR TITLE
Review changes

### PR DESCRIPTION
**Pull request by PeterChaplin**
Tcl
----
- [x] Does each proc have a #| comment to describe what it does?
- [x] Are there comments throughout the code?
- [x] Is each proc unit testable?
- [x] Are the proc names well chosen? - Needs discussion.
- [x] Are variable names well chosen?
- Security: Is user input sanitised? - API procs, should not be called with user input
- Security: Is substitution protected from injection attacks? - API procs, should not be called with user input
- [x] List what testing has been done.

### CODE REVIEW

File: tcl/image_filesystem_cache.tcl [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        lappend file_path_parts ${file_id}${ext}
+
+        set cache_file [join $file_path_parts "/"]
+
+        file mkdir [file dirname $cache_file]
+        file rename -force $file $cache_file
+    }
+}
+
+proc qc::image_filesystem_cache_smaller_biggest_exists {args} {
```
Comment (bvw) [inline tcl/image_filesystem_cache.tcl]: Would this be a better name 
`qc::image_filesystem_cache_biggest_exists`

File: tcl/image_filesystem_cache.tcl [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    #| Extract the width and height from a filesystem cache path
+    set expression {[^0-9]([0-9]+)x([0-9]+)[^0-9]}
+    regexp $expression $file -> width height
+    return [list $width $height]
+}
+
+################################################################################
+# Filesystem Cache of Original Image
+
+proc qc::image_fileststem_cache_original_exists {args} {
```
Comment (bvw) [inline tcl/image_filesystem_cache.tcl]: typo `qc::image_fileststem_cache_original_exists`
filesystem

File: tcl/image_filesystem_cache.tcl [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    set glob_pattern "autocrop/${file_id}.*"
+    set glob_options [list -nocomplain -types f -directory $cache_dir]
+    if { [llength [glob {*}$glob_options {*}$glob_pattern]] == 1 } {
+        return true
+    } else {
+        return false
+    }
+}
+
+proc qc::qc::image_fileststem_cache_autocrop_data {args} {
```
Comment (bvw) [inline tcl/image_filesystem_cache.tcl]: typo

File: tcl/image_filesystem_cache.tcl [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+    lassign [qc::image_filesystem_cache_file2dimensions $file] \
+        width height
+    set url [qc::file2url $file]
+    set timestamp [qc::cast timestamp [file mtime $file]]
+
+    return [dict_from file width height url timestamp]
+}
+
+proc qc::qc::image_fileststem_cache_autocrop_create {args} {
```
Comment (bvw) [inline tcl/image_filesystem_cache.tcl]: typo

File: tcl/image_nsv_cache.tcl [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    if { [nsv_exists image_cache_data $nsv_key] } {
+        set cache [nsv_get $nsv_key]
+    } else {
+        set cache [dict create]
+    }
+    dict set cache [list $width $height] $data
+    nsv_set image_cache_data $nsv_key $cache
+}
+
+proc qc::image_nsv_cache_smaller_biggest_exists {args} {
```
Comment (bvw) [inline tcl/image_nsv_cache.tcl]: smaller_biggest ?

**Comment by bvw** [comment]
Checklist please

File: tcl/image_filesystem_cache.tcl (hunk 1 above)
Comment (PeterChaplin) [inline tcl/image_filesystem_cache.tcl]: Not sure... it's "If the biggest (original) version of the image exists, and is smaller than (or equal to) the specified constraints".
`qc::image_filesystem_cache_smaller_original_exists`?
`qc::image_filesystem_cache_biggest_exists_and_is_smaller`?

**Comment by PeterChaplin** [comment]
Checklist added.

**Comment by PeterChaplin** [comment]
Testing (so far)
----
- local make-install
- restarted services
- Viewed image-heavy pages
- Ran image-archive proc from control-port

**Comment by PeterChaplin** [comment]
Encountered some more bugs while testing...

**Comment by PeterChaplin** [comment]
Testing
----
- Deleted cache on DEV, restarted services
- Viewed image-heavy page
- Viewed an autocropped image at smaller-than-original dimensions
- Viewed an autocropped image at larger-than-original dimensions
- Viewed an uncropped image at smaller-than-original dimensions
- Viewed an uncropped image at larger-than-original dimensions
- Built a sample image archive and downloaded

**Pull request merged**